### PR TITLE
Add Range Option

### PR DESCRIPTION
**Pull request by ghost**
Added range option while some bugs need to be fixed.

Signed-off-by: Zongyu zy728@york.ac.uk

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/Llama-Group/Project-Llama/pull/135" target="_blank">ReviewNinja

File: llama_cpp/example/data_generator/main.cpp [hunk 1]
```diff
@@ -29,9 +29,9 @@ int main(int argc, const char * argv[]) {
     std::vector<double> br, bc;
     std::vector<std::string> cr, cc;
 
-    llama::DataGenerator::generateRandomData(&ar);
-    llama::DataGenerator::generateRandomData(&br);
-    llama::DataGenerator::generateRandomData(&cr);
```
Comment (0xsuu) [inline llama_cpp/example/data_generator/main.cpp:34]: There should work as well. Why remove?

File: llama_cpp/example/data_generator/main.cpp [hunk 2]
```diff
@@ -88,7 +88,7 @@ int main(int argc, const char * argv[]) {
     }
 
     std::cout << "===== Single Int =====" << '\n';
-    std::cout << llama::DataGenerator::generateSingleDatum(&as, LE) << '\n';
```
Comment (0xsuu) [inline llama_cpp/example/data_generator/main.cpp:91]: Why remove this?

File: llama_cpp/src/Utility/DataGenerator.cpp [hunk 3]
```diff
@@ -22,42 +22,66 @@
...
 }
 
 template<>
-void DataGenerator::generateRandomData<double>(std::vector<double> *targetVector, int count) {
-    double doubleMin = std::numeric_limits<double>::min();
-    double doubleMax = std::numeric_limits<double>::max();
-    std::random_device doubleRandomDevice;
-    std::mt19937 doubleRandomEngine(doubleRandomDevice());
-    std::uniform_real_distribution<double> uniformDoubleDistribution(doubleMin, doubleMax);
+void DataGenerator::generateRandomData<double, double>(std::vector<double> *targetVector, double min, double max, int count) {
```
Comment (0xsuu) [inline llama_cpp/src/Utility/DataGenerator.cpp:68]: Unnecessary blank line.

File: llama_cpp/src/Utility/DataGenerator.cpp [hunk 4]
```diff
@@ -22,42 +22,66 @@
...
+double DataGenerator::generateRandomDataFromRange<double>(double min, double max) {
+    std::random_device doubleRandomDevice;
+    std::mt19937 doubleRandomEngine(doubleRandomDevice());
+    std::uniform_real_distribution<double> uniformDoubleDistribution(min, max);
+    return uniformDoubleDistribution(doubleRandomEngine);
+}
+
+std::string DataGenerator::generateSingleString() {
+    static std::string alpha = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+    "§1234567890-=[]';\\,./`±!@£$^&*()_+{}:\"|<>?~";
```
Comment (0xsuu) [inline llama_cpp/src/Utility/DataGenerator.cpp:42]: Should apply the alphabet we talked before. (printable ascii)

File: llama_cpp/src/Utility/DataGenerator.h [hunk 5]
```diff
@@ -45,10 +58,16 @@ class DataGenerator {
...
-    template<typename T>
-    static void generateRandomData(std::vector<T> *targetVector, int count = DATA_GENERATOR_DEFAULT_COUNT);
+    template<typename T, typename Y>
+    static void generateRandomData(std::vector<T> *targetVector,
+                                   Y min = std::numeric_limits<Y>::min(),
+                                   Y max = std::numeric_limits<Y>::max(),
+                                   int count = DATA_GENERATOR_DEFAULT_COUNT);
 
-    static void generateRandomDataFromSet(const std::string& set, std::vector<std::string> *targetVector, int count);
+    static void generateRandomDataFromSet(std::vector<std::string> *targetVector,
```
Comment (0xsuu) [inline llama_cpp/src/Utility/DataGenerator.h:67]: Why the function name is not talking about string but actually generate some strings?

**Pull request closed without merging**